### ISSUE TITLE
ERROR: The cluster's authentication mode must be set to one of [API, API_AND_CONFIG_MAP] to perform this operation.

### ISSUE THREAD

**Issue by gowgopal83** [terraform-aws-modules/terraform-aws-eks#3048]
## Description

We are trying to upgrade the EKS to `1.28` from `1.27` and the module (from 19.10.0 to 20.11.1) and also migrate the authentication mode from `Auth_config` to `API-AND-CONFIG_MAP`. I followed the upgrade v19-v20 document, upgraded the providers, and added the necessary entries. The execution is failing with the `The cluster authentication must be set` error. 

- [X ] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]:

- Terraform version: v1.3.2
- Provider version(s): 
```
Terraform v1.3.2
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v5.50.0
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.4
+ provider registry.terraform.io/hashicorp/helm v2.9.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.30.0
+ provider registry.terraform.io/hashicorp/null v3.2.2
+ provider registry.terraform.io/hashicorp/time v0.11.1
+ provider registry.terraform.io/hashicorp/tls v4.0.5
```
## Reproduction Code [Required]
```
module "eks" {
  # checkov:skip=CKV_TF_1
  source              = "terraform-aws-modules/eks/aws"
  version             = "~> 20.11.1"
  cluster_name        = local.cluster_name
  cluster_version     = var.kubernetes_version
  enable_irsa         = true
  vpc_id              = module.new_vpc.vpc_id
  subnet_ids          = module.new_vpc.private_subnets
  authentication_mode = "API_AND_CONFIG_MAP"

  cluster_endpoint_public_access           = true
  cluster_endpoint_private_access          = true
  enable_cluster_creator_admin_permissions = true

  cluster_endpoint_public_access_cidrs = concat(
    var.cidr_blocks,
    split(",", data.aws_ssm_parameter.nat_gw_eips.value)
  )

  # https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html
  cluster_enabled_log_types = ["api", "audit", "authenticator", "controllerManager", "scheduler"]

  cluster_encryption_config = {
    provider_key_arn = aws_kms_key.etcd_encryption_key.arn
    resources        = ["secrets"]
  }

  cluster_addons = {
    coredns = {
      addon_version = "v1.10.1-eksbuild.10"
      configuration_values = jsonencode({
        corefile = <<EOF
      .:53 {
          errors
          health
          hosts {
            ${local.coredns_host_config}
            fallthrough
          }
          kubernetes cluster.local in-addr.arpa ip6.arpa {
            pods insecure
            fallthrough in-addr.arpa ip6.arpa
          }
          prometheus :9153
          forward . /etc/resolv.conf
          cache 29
          loop
          reload
          loadbalance
      }
      EOF
      })
    }
    kube-proxy = {
      addon_version = "v1.28.6-eksbuild.2"
    }
    vpc-cni = {
      addon_version            = "v1.18.1-eksbuild.1"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
      configuration_values = jsonencode({
        env = {
          # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
          ENABLE_PREFIX_DELEGATION = "true"
          WARM_PREFIX_TARGET       = "1"
          WARM_IP_TARGET           = "5"
          MINIMUM_IP_TARGET        = "2"
        }
      })
    }
    # AWS Distro for OpenTelemetry
    adot = {
      addon_version = "v0.94.1-eksbuild.1"
      # https://github.com/terraform-aws-modules/terraform-aws-eks/issues/2731
      most_recent = true
    }
    aws-efs-csi-driver = {
      addon_version            = "v2.0.1-eksbuild.1"
      service_account_role_arn = module.efs_csi_irsa_role.iam_role_arn
    }
  }

  iam_role_additional_policies = {
    additional = var.cloudwatch_server_policy_arn
  }

  node_security_group_additional_rules = {
    ingress_self_https = {
      description = "Enable https self traffic"
      protocol    = "TCP"
      from_port   = 443
      to_port     = 443
      type        = "ingress"
      self        = true
    }
    ingress_self_http = {
      description = "Enable http self traffic"
      protocol    = "TCP"
      from_port   = 80
      to_port     = 80
      type        = "ingress"
      self        = true
    }
  }

  # As we have enabled `attach_cluster_primary_security_group = true` we had to set the below tag as null
  node_security_group_tags = {
    "kubernetes.io/cluster/${local.cluster_name}" = null
  }

nodegroup configs
  eks_managed_node_group_defaults = {
    instance_types                        = var.managed_node_group_instance_type
    iam_role_attach_cni_policy            = true
    attach_cluster_primary_security_group = true # default is false
    iam_role_additional_policies = {
      eksKedaCloudWatchMetricsHPA = aws_iam_policy.keda_cloudwatch_metrics.arn
      additional                  = data.aws_iam_policy.cw_agent_policy.arn
      ssmPolicy                   = data.aws_iam_policy.ssm_agent_policy.arn # Added ssm policy to the nodes to login to the nodes
    }
    pre_bootstrap_user_data = <<-EOT
      export USE_MAX_PODS=false
      export KUBELET_EXTRA_ARGS="--max-pods=110"
    EOT
  }

  eks_managed_node_groups = {
    node-group-1 = {
      min_size     = local.eks_asg_min_node_count
      max_size     = local.eks_asg_max_node_count
      desired_size = local.eks_asg_desired_node_count

      instance_types = var.managed_node_group_instance_type
      capacity_type  = var.eks_capacity_type
      ami_id         = var.eks_worker_ami
      cpu_credits    = "unlimited"
      enable_bootstrap_user_data = true

      tags = local.cluster_autoscaler_tags
    }
  }

  fargate_profiles = {
    argo-events = {
      name = "argo-events"
      selectors = [
        {
          namespace = "argo-events"
          labels = {
            Application = "argo-fargate"
          }
        }
      ]
      tags = {
        Owner = "secondary"
      }
    }
  }


  tags = local.tags
}

module "eks_aws_auth" {
  # checkov:skip=CKV_TF_1
  source                    = "terraform-aws-modules/eks/aws//modules/aws-auth"
  version                   = "~> 20.11.1"
  manage_aws_auth_configmap = true
  aws_auth_roles            = local.eks_iam_rbac_mappings
}
```

 yes 
 yes 
 Upgrading the module from `19.10.0` to `20.11.1`

## Expected behavior

EKS upgrade and the authentication_mode migrated to `API_AND_CONFIG_MAP`

## Actual behavior
EKS is upgraded to `1.28` however the code failed to create the nodes and failed with the following error.

```
 Error: creating EKS Access Entry (aws_core_eks:arn:aws:iam::681496624581:role/admin): operation error EKS: CreateAccessEntry, https response error StatusCode: 400, RequestID: c77dd87a-bdc7-41fb-b1ba-b45027c03766, InvalidRequestException: The cluster's authentication mode must be set to one of [API, API_AND_CONFIG_MAP] to perform this operation.
│
│   with module.eks.aws_eks_access_entry.this["cluster_creator"],
│   on .terraform/modules/eks/main.tf line 191, in resource "aws_eks_access_entry" "this":
│  191: resource "aws_eks_access_entry" "this" {
│
╵
```

## Additional context
Note:
Have performed the following state removal before applying the changes. 
```
terraform state rm 'module.eks.kubernetes_config_map_v1_data.aws_auth[0]'
```

 State `module.eks.kubernetes_config_map.aws_auth[0]` was not present.

**Comment by gowgopal83** [comment]
Have downgraded the module version to 20.8.2 which solved the above issue and upgraded the EKS cluster.

**Comment by gowgopal83** [comment]
> Have downgraded the module version to 20.8.2 which solved the above issue and upgraded the EKS cluster.

After moving to `authentication_mode=API`, have upgraded to the latest tag.

**Comment by yurrriq** [comment]
We had the same issue and tried upgrading to k8s first 1.28 separately (with module version 19) and got the same errors even with pinning 20.8.2 in a subsequent change.

**Comment by yurrriq** [comment]
In the end we had to set `enable_cluster_creator_admin_permissions = false` and run `terraform apply` twice, since there appears to be a race condition (at least in v20.8.2 and v20.16.0) between changing the `authentication_mode` and `module.karpenter.aws_eks_access_entry.node[0]`. I found https://github.com/terraform-aws-modules/terraform-aws-eks/blob/v20.16.0/modules/karpenter/main.tf#L685 but clearly that's insufficient. Can we re-open this ticket? Or should I open a new one?

A (very) minimal example for the Karpenter issue is something like the following.

```terraform
module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 20.0" # i.e. 20.8.2 or 20.16.0
  # ...
  # NOTE: was "CONFIG_MAP"
  authentication_mode = "API_AND_CONFIG_MAP"
  # ...
  # NOTE: was "1.27" but this part proved irrelevant in testing
  cluster_version = "1.28"
  # ...
}

module "karpenter" {
  source  = "terraform-aws-modules/eks/aws//modules/karpenter"
  version = "~> 20.0"
  # ...
}
```

**Comment by bryantbiggs** [comment]
> since there appears to be a race condition (at least in v20.8.2 and v20.16.0) between changing the authentication_mode and module.karpenter.aws_eks_access_entry.node[0]

correct - changes to the access entry API mode are not immediate so if you try to change the mode and use auth based on that mode, there is a race condition. its recommended to change the access entry mode first, then make any downstream changes

> N.B. We have not tried setting enable_cluster_creator_admin_permissions = true after successfully upgrading to 20.x and setting authentication_mode = "API_AND_CONFIG_MAP" but I assume that will result in a simliar error to the one we've seen:

This usually occurs with folks trying to create access entries for the managed node group IAM role and/or the cluster creator - you can see more details [here](
https://github.com/terraform-aws-modules/terraform-aws-eks/issues/2968#issuecomment-1991863351)

tl;dr - when converting from configmap only (default case for all clusters created prior to access entry), EKS will automatically add an access entry for any roles that are used by EKS managed node group, Fargate profiles, as well as the cluster creator identity. Our version via `enable_cluster_creator_admin_permissions` is not the same as the EKS API version - which is a one time decision at time of cluster creation. Our implementation here is using the access entry API to allow full control over those changes - now and in the future. (ref https://github.com/terraform-aws-modules/terraform-aws-eks/blob/c52bbb32a41e5f5cd2b1fd1d9b04ee1baa8e496e/main.tf#L38-L43)

**Comment by yurrriq** [comment]
Thanks very much for this explanation! Totally clears it up.

**Comment by bryantbiggs** [comment]
also - there are new docs coming that I hope will help with some of these questions - we have been woefully behind in terms of providing these details and nuances so thats on me to get those out https://github.com/terraform-aws-modules/terraform-aws-eks/compare/master...bryantbiggs:terraform-aws-eks:chore/new-docs
